src/classes/PS_IntegrationServiceConfig.cls

```diff
@@ -66,12 +90,86 @@ public with sharing class PS_IntegrationServiceConfig {
     * @return Boolean
     */
     public Boolean hasIntegrationPermissions() {
-        // If the integration service is not enabled, user does not have the integration permissions
-        if (!isEnabled) {
-            return false;
+        Payment_Services_Configuration__c lastModifiedRecord = new Configuration().lastModifiedRecord;
+        return lastModifiedRecord != null 
+                ? lastModifiedRecord.LastModifiedById == UserInfo.getUserId() 
+                : false;
+    }
+
+
+
+    /**
+    * @description Inner class responsible for retrieving Payment Services configuration from
+    *              application storage and building and exposing a set of key-value
+    *              configuration pairs.
+    *              We use 'without sharing' to ensure that the query against 
+    *              Payment_Services_Configuration__c always works regardless of what permissions 
+    *              the current user has
+    */
+    @TestVisible
+    public without sharing class Configuration {
+        public Configuration() {}
+
+        public Map<String, String> keyValueMap {
+            get {
+                if (keyValueMap == null) {
+                    keyValueMap = new Map<String, String>();
+                    String makanaKey;
+                    for (Payment_Services_Configuration__c configRecord :
+                            getConfigRecordsByName(
+                                    new List<String>{
+                                            PAYMENTS_SERVICE_NAME, MAKANA_SERVICE_NAME
+                                    }
+                            )) {
+                        if (configRecord.Service__c == PAYMENTS_SERVICE_NAME) {
+                            keyValueMap.put(configRecord.Key__c, configRecord.Value__c);
+                        } else if (configRecord.Key__c == API_KEY &&
+                                configRecord.Service__c == MAKANA_SERVICE_NAME) {
+                            makanaKey = configRecord.Value__c;
+                        }
+                    }
+
+                    // If the payments service config does not have an Api key
+                    // and makana does, use the makana Api key
+                    if (makanaKey != null && keyValueMap.get(API_KEY) == null) {
+                        keyValueMap.put(API_KEY, makanaKey);
+                    }
+                }
+                return keyValueMap;
+            }
+            private set;
+        }
+
+        public Payment_Services_Configuration__c lastModifiedRecord {
+            get {
+                if (lastModifiedRecord == null) {
+                    List<Payment_Services_Configuration__c> lastModifiedRecords = getLastModifiedConfigRecord();
```

Comment:
For this method and `getConfigRecordsByName`, Is the intent that every time this method is called it re-queries the object, even in the same execution context? I'm wondering if it makes sense to store the result of the query into a class var and allow that results to be re-used on all subsequent calls into this method/class - the goal being to reduce the number of queries. I can see where that might not be advantageous if the data could change mid-context, but if I understand the usage correctly I think that is not likely.

Reply:
We are checking `lastModifiedRecord == null`. Wouldn't that prevent the query execution for the second time?

Reply:
Excellent point. The getter will only execute the first time.